pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -125,16 +130,15 @@ public boolean isFinal() {
   private final Schema _schema;
   private final ServerMetrics _serverMetrics;
   private final RealtimeSegmentImpl _realtimeSegment;
-  private long _endOffset = Long.MAX_VALUE; // No upper limit on Kafka offset
-  private long _currentOffset;
+  private volatile long _currentOffset;
   private volatile State _state;
-  private int _numRowsConsumed = 0;
+  private volatile int _numRowsConsumed = 0;
```

Comment:
We shouldn't need the volatile, unless the consumption loop runs concurrently with catching up. If it does, we should prevent that from happening instead of trying to synchronize between both threads.